src/ListView.vala

```diff
@@ -147,78 +183,93 @@ public class Tasks.ListView : Gtk.Grid {
-        objects.foreach ((ical_comp) => {
-            try {
-                SList<ECal.Component> ecal_tasks;
-                client.get_objects_for_uid_sync (ical_comp.get_uid (), out ecal_tasks, null);
-
-                ecal_tasks.foreach ((task) => {
-                    debug_task (source, task);
-
-                    if (!added_tasks.contains (task)) {
-                        added_tasks.add (task);
-                    }
-                });
+    [CCode (instance_pos = -1)]
+    private int sort_function (Gtk.ListBoxRow row1, Gtk.ListBoxRow row2) {
+        var row_a = (Tasks.TaskRow) row1;
+        var row_b = (Tasks.TaskRow) row2;
+
+        if (row_a.completed == row_b.completed) {
+            unowned ICal.Component comp_a = row_a.task.get_icalcomponent ();
+            unowned ICal.Component comp_b = row_b.task.get_icalcomponent ();
+
+            ICal.Time start_a = comp_a.get_dtstart ();
+            ICal.Time stamp_a = comp_a.get_dtstamp ();
+
+            ICal.Time start_b = comp_b.get_dtstart ();
+            ICal.Time stamp_b = comp_b.get_dtstamp ();
+
+            if ( start_a.is_null_time () && start_b.is_null_time () ) {
+                return stamp_b.compare (stamp_a);
 
-            } catch (Error e) {
-                warning (e.message);
+            } else if (start_a.is_null_time () && !start_b.is_null_time ()) {
+                return 1;
+
+            } else if (start_b.is_null_time () && !start_a.is_null_time ()) {
+                return -1;
+
+            } else {
+                return start_a.compare (start_b);
             }
-        });
 
-        tasks_added (client, source, added_tasks.read_only_view);
+        } else if (row_a.completed && !row_b.completed) {
+            return 1;
+
+        } else if (row_b.completed && !row_a.completed) {
+            return -1;
+        }
+
+        return 0;
     }
 
-    private void tasks_added (ECal.Client client, E.Source source, Gee.Collection<ECal.Component> tasks) {
+    private void header_function (Gtk.ListBoxRow lbrow, Gtk.ListBoxRow? lbbefore) {
+        if (source != null || !(lbrow is Tasks.TaskRow)) {
+            return;
+        }
+        var row = (Tasks.TaskRow) lbrow;
+        unowned ICal.Component comp = row.task.get_icalcomponent ();
+
+        if (comp.get_due ().is_null_time ()) {
+            return;
+        }
+
+        if (lbbefore != null) {
+            var before = (Tasks.TaskRow) lbbefore;
+            unowned ICal.Component comp_before = before.task.get_icalcomponent ();
+
+            if (comp_before.get_due ().compare_date_only (comp.get_due ()) == 0) {
+                return;
+            }
+        }
+
+        var due_date_time = Util.ical_to_date_time (comp.get_due ());
+        var header_label = new Granite.HeaderLabel (Tasks.Util.get_relative_date (due_date_time));
+        header_label.ellipsize = Pango.EllipsizeMode.MIDDLE;
+        header_label.margin_left = 4;
```

Comment:
`margin_left` is deprecated. Also doesn't align with the editing card. Should be:
```suggestion
        header_label.margin_start = 6;
```